### PR TITLE
Feature for render callback of stack screen

### LINKED ISSUE #31: Binding for Stack-component render-callback .
## Feature request

https://reactnavigation.org/docs/hello-react-navigation#passing-additional-props

Stack component can accept render callback instead of component prop.

eg:
```js
<Stack.Screen name="Home">
  {props => <HomeScreen {...props} extraData={someData} />}
</Stack.Screen>
```

### PR DESCRIPTION
**Pull request by a-c-sreedhar-reddy**
Closes #31

### CODE REVIEW

**Comment by MoOx** [comment]
This introduce a cost for the binding and a door to bugs.
Why not keeping a simple binding with 2 props like you mentioned here https://github.com/reason-react-native/reason-react-navigation/issues/31#issuecomment-637265149 ? Maybe the binding should not be responsible for checking that you use the lib correctly...

**Comment by a-c-sreedhar-reddy** [comment]
But if with that both of these can be None or both of them can be provided, which would lead to app crash.

**Comment by MoOx** [comment]
react navigation really trigger a crash if you forget component or children?

**Comment by a-c-sreedhar-reddy** [comment]
Yeah I tried it

**Comment by a-c-sreedhar-reddy** [comment]
WhatsApp Image 2020-06-02 at 2 17 43 PM
WhatsApp Image 2020-06-02 at 2 17 43 PM (1)

**Comment by MoOx** [comment]
If I am correct this error happen on first launch, when the routing is being processed right? Isn't it acceptable to have this "runtime error" with zero cost bindings?

**Comment by MoOx** [comment]
I am thinking about another zero cost solution:

implement 2 components Screen (component) & ScreenWithCallback (render prop)

**Comment by a-c-sreedhar-reddy** [comment]
Implemented two components Screen and ScreenWithCallback.

File: src/Stack.re [hunk 1]
```diff
@@ -266,17 +262,23 @@ module Make = (M: {type params;}) => {
...
-        "route": route,
-      }),
+      option(
+        React.component({
+          .
+          "navigation": navigation,
+          "route": route,
+        }),
+      ),
+    children: option(renderCallbackProp => React.element),
```
Comment (MoOx) [inline src/Stack.re:281]: Maybe instead we should have screenProps & screenPropsWithCallback?

Comment (a-c-sreedhar-reddy) [reply src/Stack.re:281]: https://github.com/reason-react-native/reason-react-navigation/blob/master/src/Stack.re#L288

But Screen here can accept only one type right?

**Comment by Freddy03h** [comment]
I'm really interested on this change!
But, maybe we should do the same change for (Bottom/Material)Tab Screen?

**Comment by MoOx** [comment]
@Freddy03h Would make sense indeed. @a-c-sreedhar-reddy can you handle this?

**Comment by MoOx** [comment]
@a-c-sreedhar-reddy up about the request :) Do you want us to handle it if you don't have time?

**Comment by a-c-sreedhar-reddy** [comment]
Yes @MoOx .🙂

**Comment by MoOx** [comment]
@Freddy03h would you like to handle that in another PR?

**Comment by Freddy03h** [comment]
The all change or the tab screen part?

**Comment by Freddy03h** [comment]
Ok I'll try to duplicate the changes on the tabs navigators

**Comment by MoOx** [comment]
Thanks!

**Pull request merged**